_posts/2022-08-22-fan-out-and-back-in-using-functions.md

```diff
@@ -0,0 +1,415 @@
+---
+title: "Exploring the Fan out and Fan in pattern with OpenFaaS"
+description: "Learn how to fan out requests to OpenFaaS functions to process them in parallel at scale, before consolidating the results by fanning back in again."
+date: 2022-08-22
+image: /images/2022-fan-out-and-back-in-using-functions/background.jpg
+categories:
+- architecture
+- dataprocessing
+- bigdata
+author_staff_member: han
+---
+
+Learn how to fan out requests to OpenFaaS functions to process them in parallel at scale, before consolidating the results by fanning back in again.
+
+We're increasingly hearing that OpenFaaS functions are convenient for ingesting, transforming and processing large amounts of data. Functions are a natural fit for processing data because they're short-lived, stateless, efficient and can scale out to processing large amounts of data in parallel.
+
+A common pattern used when processing large amounts of data is to break down an expensive task into smaller sub tasks that can be executed in parallel. This fan-out pattern is already available in OpenFaaS through asynchronous functions. With the async feature, invocations are queued up and processed as and when there's capacity.
+
+> Our post, [How to process your data the resilient way with back pressure ](https://www.openfaas.com/blog/limits-and-backpressure/), shows how OpenFaaS can be used to process a dataset as quickly as possible, without losing any of the records.
+
+In some cases we want to be notified when a group of asynchronous function invocations completes and collect the result from each individual invocation. We need to fan-in again. Fanning in requires a bit more code. We will show how this can be implemented with OpenFaaS functions.
```

Comment:
The MapReduce pattern was popularised by Hadoop, which specialises in processing large batches of data. One dataset or task is split into many others, which can be worked on in parallel over a large distributed network of computers, that's the map step. The reduce step is where the results are combined to make sense of them.

We've seen customers making use of OpenFaaS in this way, so we wanted to show you how we think it can be done without making any changes to OpenFaaS itself.